First commit

peachblacky opened the pull request.

peachblacky committed:
First commit

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+import java.util.Objects;
+import java.util.StringTokenizer;
+
+public class Server implements Runnable{
```
принципы SOLID, буква S - single responsibility
Server implements Runnable это нарушает, у тебя серверу не надо быть раннабл, раннабл - то что обрабатывает полученный сервером сокет

Gregpack replied:
надо разделить логику сервера, мейн и обработчика сокета в разные классы-файлы

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+        connect = c;
+    }
+
+    public static void main(String[] args) {
```
в мэйн классе лучше чтобы ничего не было кроме main метода и вызова какой-нибудь одной функции в нём

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+                StringTokenizer parse = new StringTokenizer(input);
+                String method = parse.nextToken().toUpperCase();
+                fileRequest = parse.nextToken().toLowerCase();
+                Hashtable<String, String> HeadersTable = getHeadersTable(in);
```
Hashtable - слишком жирно, HashMap хватит. хэштейбл, во-первых, оч старая херня, во вторых несет в себе встроенные синхронизации для многопоточности

Gregpack replied:
https://stackoverflow.com/questions/40471/what-are-the-differences-between-a-hashmap-and-a-hashtable-in-java

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    private Hashtable<String, String> getHeadersTable(BufferedReader in) throws IOException {
+        Hashtable<String, String> HeadersTable = new Hashtable<>();
```
hashtable

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+        while(true) {
+            System.out.println("ITERATION");
+            try {
+                this.getConnect().setSoTimeout(20000);
```
зачем? а если чел использующий браузер отошел на 30 секунд то чё теперь, умереть?

peachblacky replied:
Типо вообще таймаут убрать? Или сделать его больше?

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+
+                    dataOut.write(fileData, 0, length);
+                    dataOut.flush();
+
```
зачем ты пишешь в один и тот же выходной поток двумя разными объектами?

peachblacky replied:
Ну тут ведь разные типы: один PrintWriter, а другой BufferedOutputStream
Через один хедеры пишу, через другой данные.

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+                    ioException.printStackTrace();
+                }
+            } catch (IOException e) {
+                System.err.println("Server error : " + e);
```
после ioexception сокет можно на помойку отправлять, а не читать дальше

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+                dataOut = new BufferedOutputStream(connect.getOutputStream());
+
+                String input = in.readLine();
+                if(input.length() == 0) {
```
NPE при input == null

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+        PrintWriter out = null;
+        BufferedOutputStream dataOut = null;
+        String fileRequest = null;
+        while(true) {
```
не хватает break-ов в эксепшенах еще, а то получается что ошибку поймали а сокетом дальше пользуемся

Gregpack commented on src/Server.java:
```diff
@@ -0,0 +1,214 @@
+                        break;
+                    }
+                    out.println("Content-type: " + content);
+                    out.println("Content-length: " + length);
```
вкусовщина, но раз уж мы по джаве идём: ты много где пишешь один и тот же код про создание HTTP ответа и печать его в поток. стоит выделить HTTP ответ в какой-нибудь класс, прикрепить к нему паттерн builder (почитай, это довольно простая штука) или просто хэшмапу заголовков, сделать в нём метод toString и получить легкорасширяемый ответ на случай если бизнес попросит добавить еще 200 заголовков сверху
короче, код расширяемый будет и красивый

peachblacky committed:
Corrected all issues, exept one...

Gregpack commented on src/QueryHandler.java:
```diff
@@ -0,0 +1,197 @@
+                    out.println("Last Modified: " + file.lastModified());
+                    out.println("Content-type: " + contMimeType);
+                    out.println("Content-length: " + length);
+                    out.flush();
```
я всё еще считаю что эту гору принтлнов которая повторяется три раза можно либо вынести в функцию либо в класс

Gregpack commented on src/QueryHandler.java:
```diff
@@ -0,0 +1,197 @@
+                    out.println("Content-type: " + content);
+                    out.println("Content-length: " + length);
+                    out.println();
+                    out.flush();
```
гора принтлнов

Gregpack commented on src/QueryHandler.java:
```diff
@@ -0,0 +1,197 @@
+        String contMimeType = "text/html";
+        byte[] fileData = readFileData(file, length);
+
+        out.println("HTTP/1.1 404 File Not Found");
+        out.println("Server : Rostiks server : 1.0");
+        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.RFC_1123_DATE_TIME;
+        ZonedDateTime date = ZonedDateTime.now(ZoneOffset.UTC);
+        out.println("Date" + date.format(dateTimeFormatter));
+        out.println("Last Modified: " + file.lastModified());
+        out.println("Content-type: " + contMimeType);
+        out.println("Content-length: " + length);
+        out.println();
+        out.flush();
```
гора принтлнов

peachblacky committed:
--Написал Builder для создания ответов
--Добавил условие выхода из цикла while: при закрытии сокета это будет происходить
